File: AIペアプログラミング実践編_ソースコード/practical-ts/session-5/exercise/src/services/productService.ts
import type { Product, ProductInput, ProductUpdateInput } from '../models'
import type { ProductRepository } from '../repositories/productRepository'
import { validateProductInput } from './validation'

/**
 * 商品サービス
 * 商品に関するビジネスロジックを担当
 */
export class ProductService {
  constructor(private repository: ProductRepository) {}

  /**
   * 商品を新規登録
   */
  async create(input: ProductInput): Promise<Product> {
    // バリデーション
    const validation = validateProductInput(input)
    if (!validation.isValid) {
      throw new Error(validation.errors.join(', '))
    }

    return this.repository.save(input)
  }

  /**
   * 全商品を取得
   */
  async findAll(): Promise<Product[]> {
    return this.repository.findAll()
  }

  /**
   * IDで商品を取得
   */
  async findById(id: number): Promise<Product | null> {
    return this.repository.findById(id)
  }

  /**
   * 商品を更新
   */
  async update(id: number, input: ProductUpdateInput): Promise<Product> {
    // 存在確認
    const existing = await this.repository.findById(id)
    if (!existing) {
      throw new Error('商品が見つかりません')
    }

    // 部分的なバリデーション
    if (input.name !== undefined) {
      if (input.name.trim().length === 0) {
        throw new Error('商品名は必須です')
      }
      if (input.name.trim().length > 100) {
        throw new Error('商品名は100文字以内にしてください')
      }
    }

    if (input.price !== undefined) {
      if (input.price < 0) {
        throw new Error('価格は0以上にしてください')
      }
      if (!Number.isInteger(input.price)) {
        throw new Error('価格は整数で指定してください')
      }
    }

    return this.repository.update(id, input)
  }

  /**
   * 商品を削除
   */
  async delete(id: number): Promise<void> {
    // 存在確認
    const existing = await this.repository.findById(id)
    if (!existing) {
      throw new Error('商品が見つかりません')
    }

    await this.repository.delete(id)
  }

  /**
   * 商品名で検索
   */
  async search(query: string): Promise<Product[]> {
    if (!query || query.trim().length === 0) {
      return this.repository.findAll()
    }
    return this.repository.search(query.trim())
  }
}

File: AIペアプログラミング実践編_ソースコード/practical-java/session-3/Ans/src/main/java/com/example/inventory/service/ProductService.java
package com.example.inventory.service;

import com.example.inventory.dto.ProductInput;
import com.example.inventory.dto.ValidationResult;
import com.example.inventory.entity.Product;
import com.example.inventory.repository.ProductRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;

/**
 * 商品サービス
 */
@Service
public class ProductService {
    private final ProductRepository productRepository;
    private final ProductValidator productValidator;

    @Autowired
    public ProductService(ProductRepository productRepository) {
        this.productRepository = productRepository;
        this.productValidator = new ProductValidator();
    }

    // テスト用コンストラクタ
    public ProductService(ProductRepository productRepository, ProductValidator productValidator) {
        this.productRepository = productRepository;
        this.productValidator = productValidator;
    }

    public Product create(ProductInput input) {
        // バリデーション
        ValidationResult validation = productValidator.validate(input);
        if (!validation.isValid()) {
            throw new IllegalArgumentException(String.join(", ", validation.getErrors()));
        }

        // エンティティの作成
        Product product = new Product();
        product.setName(input.getName());
        product.setCategory(input.getCategory());
        product.setPrice(input.getPrice());

        return productRepository.save(product);
    }

    public List<Product> findAll() {
        return productRepository.findAll();
    }

    public Product findById(Long id) {
        return productRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("商品が見つかりません: " + id));
    }

    public List<Product> search(String name) {
        return productRepository.findByNameContaining(name);
    }
}
